_posts/2016-03-28-Mutual-Problems.md

```diff
@@ -0,0 +1,130 @@
+
+From the file in the `PKCS12` format we can generate another keystore by importing our `PKCS12` into the new keystore:
+
+    $ keytool -importkeystore \
+        -destkeystore key.jks \
+        -deststorepass <<keystore_password>> \
+        -destkeypass <<key_password_in_keystore>> \
+        -alias client \
+        -srckeystore key.p12 \
+        -srcstoretype PKCS12 \
+        -srcstorepass <<original_password_of_PKCS12_file>>
+
+This command looks a little bit more complex, but it is fairly easy to decrypt. At the beginning of the command we declare the parameters of the new keystore named `key.jks`. We define password for the keystore and the password for the private key which will be used by this keystore. We also assign the private key to some alias in the keystore (in this case it is `client`).
+Next we specify the source file (`key.p12`), the format of this file and the original password.
+
+With `trusted.jks` and `key.jks` we were ready for coding. In the first step we had to describe how we want to use our keystores. 
+
+```java
+File trustedKeystoreFile = new File("trusted.jks");
+File keystoreFile = new File("key.jks");
+
+SSLContext sslcontext = SSLContexts.custom()
+    .loadTrustMaterial(trustedKeystoreFile, 
+                    "<<trusted_keystore_password>>".toCharArray())
+    .loadKeyMaterial(keystoreFile, 
+                    "<<keystore_password>>".toCharArray(), 
+                    "<<original_password_of_PKCS12_file>>".toCharArray())
+    .build();
+
+SSLConnectionSocketFactory sslSocketFactory = new SSLConnectionSocketFactory(
+                sslcontext,
+                new String[]{"TLSv1.2"},
+                null,
+                SSLConnectionSocketFactory.getDefaultHostnameVerifier());
+```
+
+We took our keystore files and we built an SSL context. Next we created the socket factory which provides proper HTTPS connection for our requests.
+
+And finally we where able to call our endpoint from **Java**:
+
+```java
+try (CloseableHttpClient httpclient = HttpClients.custom()
+        .setSSLSocketFactory(sslsf)
+        .build()) {
+
+    HttpGet httpGet = new HttpGet("https://ourserver.com/our/endpoint");
+
+    try (CloseableHttpResponse response = httpclient.execute(httGet)) {
+        HttpEntity entity = response.getEntity();
+        System.out.println(response.getStatusLine());
+        EntityUtils.consume(entity);
+    }
+}
+```
+
+Done. After creating two additional files (keystores) which were the equivalent of our original certificate and the private key we implemented **the mutual authentication** with **Java**. Maybe the implementation of HTTPS connections in **Java** had some justification, but now it is just a headache. 
```

Comment:
we implemented **mutual authentication** in **Java**

ha`s` some justification